src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -77,62 +89,77 @@ void WriteBufferFromAzureBlobStorage::execWithRetry(std::function<void()> func,
 
 void WriteBufferFromAzureBlobStorage::finalizeImpl()
 {
+    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
+
+    if (buffer_allocation_policy->getBufferNumber() == 1)
+    {
+        size_t data_size = size_t(position() - memory.data());
+        if (data_size <= max_single_part_upload_size)
+        {
+            Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(memory.data()), data_size);
+            execWithRetry([&](){ block_blob_client.Upload(memory_stream); }, max_unexpected_write_error_retries, data_size);
+            LOG_TRACE(log, "Committed single block for blob `{}`", blob_path);
+            return;
+        }
+    }
+
+
     execWithRetry([this](){ next(); }, max_unexpected_write_error_retries);
 
-    if (tmp_buffer_write_offset > 0)
-        uploadBlock(tmp_buffer->data(), tmp_buffer_write_offset);
+    task_tracker->waitAll();
 
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
     execWithRetry([&](){ block_blob_client.CommitBlockList(block_ids); }, max_unexpected_write_error_retries);
 
     LOG_TRACE(log, "Committed {} blocks for blob `{}`", block_ids.size(), blob_path);
 }
 
-void WriteBufferFromAzureBlobStorage::uploadBlock(const char * data, size_t size)
+void WriteBufferFromAzureBlobStorage::nextImpl()
 {
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
-    const std::string & block_id = block_ids.emplace_back(getRandomASCIIString(64));
-
-    Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(data), size);
-    execWithRetry([&](){ block_blob_client.StageBlock(block_id, memory_stream); }, max_unexpected_write_error_retries, size);
-    tmp_buffer_write_offset = 0;
-
-    LOG_TRACE(log, "Staged block (id: {}) of size {} (blob path: {}).", block_id, size, blob_path);
+    task_tracker->waitIfAny();
+    writePart();
+    allocateBuffer();
 }
 
-WriteBufferFromAzureBlobStorage::MemoryBufferPtr WriteBufferFromAzureBlobStorage::allocateBuffer() const
+void WriteBufferFromAzureBlobStorage::allocateBuffer()
 {
-    return std::make_unique<Memory<>>(max_single_part_upload_size);
+    buffer_allocation_policy->nextBuffer();
+    auto size = buffer_allocation_policy->getBufferSize();
+
+    if (buffer_allocation_policy->getBufferNumber() == 1)
+    {
+        size = std::min(size_t(DBMS_DEFAULT_BUFFER_SIZE), size);
```

Comment:
It looks really strange that we have a cool separate class for calculating buffer size and immediately after we call it we have to immediately patch the result. `BufferAllocationPolicy` already has a check for `min_size`, doesn't it?

Reply:
ya, moved this to BufferAllcationPolicy & removed it from S3 & AzureBlobStorage WriteBuffers.